Add docs for disconnected block (#20147)
hashicorp/nomad: website/content/docs/job-specification/disconnect.mdx

---
layout: docs
page_title: disconnect Block - Job Specification
description: |-
  The "disconnect" block describes the behavior of both the Nomad server and 
  client in case of a network partition, as well as how to reconcile the workloads
  in case of a reconnection.
---

# `disconnect` Block

<Placement groups={['job', 'group', 'disconnect']} />

> The `disconnect` block describes the system's behavior in case of a network 
> partition.

Reviewer: Let's put this right at the very top, so it's clear what this block is for.

```suggestion
The `disconnect` block describes the system's behavior in case of a network 
partition. By default, without a `disconnect` block, if an allocation is on a 
node that misses heartbeats, the allocation will be marked `lost` and will be 
rescheduled.
```

---
layout: docs
page_title: disconnect Block - Job Specification
description: |-

>   The "disconnect" block describes the behaviuor of both the Nomad server and 

Reviewer: ```suggestion
  The "disconnect" block describes the behavior of both the Nomad server and 
```

  client in case of a network partition, as well as how to reconcile the workloads
  in case of a reconnection.
---

# `disconnect` Block

<Placement groups={['job', 'group', 'disconnect']} />

> The `disconnect` block describes the systm's behaviour in case of a network 

Reviewer: ```suggestion
The `disconnect` block describes the system's behavior in case of a network 
```

partition.

```hcl
    job "docs" {
  	    group "example" {
    	    disconnect {
		        lost_after = "6h"
		        stop_after = "2h"
		        replace = false
                reconcile = "keep_original"
     	    }
  	    }
    }

```

## `disconnect` Parameters

- `lost_after` `(string: "")` - Specifies a duration during which a Nomad client
  will attempt to reconnect allocations after it fails to heartbeat

>   in the [`heartbeat_grace`] window.  It defaults to “” which is equivalent to 

Reviewer: Some syntax highlighting of markdown will only pick up this as a link if you have a trailing empty bracket here. It's not required for rendering, but it's a nice-to-have:

```suggestion
  in the [`heartbeat_grace`][] window.  It defaults to "" which is equivalent to 
```

  having the disconnect block be nil.
  
  See [the example code below][lost_after] for more details. This setting cannot
  be used with [`stop_after`].

- `replace` `(bool: false)` - Specifies if the disconnected allocation should 
  be replaced by a new one rescheduled on a different node. If false and the 

>   node it is running on becomes disconnected or goes down, this allocations
>   wont be rescheduled and will show up as `unknown` until the node comes back 
>   up or it are manually stoped:

Reviewer: ```suggestion
  node it is running on becomes disconnected or goes down, this allocation
  won't be rescheduled and will be reported as `unknown` until the node reconnects, 
  or until the allocation is manually stopped:
```

(I think that's what we mean in that last part?)

  ```plaintext
  `nomad alloc stop  <alloc ID>`
  ```

>   If true, a new  alloc will be placed immediately upon the node becoming 

Reviewer: ```suggestion
  If true, a new alloc will be placed immediately upon the node becoming 
```

  disconnected.

- `stop_after` `(string: "")` - Specifies a duration after which a disconnected 

>   Nomad client will stop its allocations. Keep in mind that if stop_after is 
>   shorter than ttl, and `replace` = false, can lead to a state where no allocations
>   are running, hence it is not admitted and will cause a validation error.

Reviewer: We're describing the "why" before the "what" here. I'd reword this to make it clear at the start of the sentence that the configuration described isn't allowed, and _then_ describe why. Also, I'm not sure what "ttl" we're referring to here?

```suggestion
  Nomad client will stop its allocations. Setting `stop_after` shorter than 
  ??? and `replace = false` at the same time is not permitted and will cause a 
  validation error, because this would lead to a state where no allocations 
  can be scheduled.
```

  The Nomad client process must be running for this to occur. This setting
  cannot be used with [`lost_after`].

> - `reconcile` `(string: “best_score”)` - Specifies which allocation to keep once

Reviewer: We probably don't want "smart quotes" anywhere, as I'm not sure the HCL parser will accept them if someone copy-pastes:

```suggestion
- `reconcile` `(string: "best_score")` - Specifies which allocation to keep once
```

  the previously disconnected node regains connectivity.
  It has four possible values which are described below:

>     - `keep_original`: Always keep the original allocation, bare in mind 

Reviewer: ```suggestion
    - `keep_original`: Always keep the original allocation. Bear in mind 
```

not be stopped on the client, regardless of how long the client is out
of touch.

Note that if the server's clocks are not closely synchronized with
each other, the server may reschedule the group before the client has
stopped the allocation. Operators should ensure that clock drift
between servers is as small as possible.

Note also that a group using this feature will be stopped on the
client if the Nomad server cluster fails, since the client will be
unable to contact any server in that case. Groups opting in to this
feature are therefore exposed to an additional runtime dependency and
potential point of failure.

```hcl
group "first" {
  stop_after_client_disconnect = "90s"

  task "first-task" {
    kill_timeout = "15s"
  }
}

group "second" {

  task "second-task" {
    kill_timeout = "5s"
  }
}
```

### Lost After

`lost_after` specifies a duration during which a Nomad client will
attempt to reconnect allocations after it fails to heartbeat in the
[`heartbeat_grace`] window.

By default, allocations running on a client that fails to heartbeat will be
marked "lost". When a client reconnects, its allocations, which may still be

> healthy, will restart because they have been marked "lost". This can cause

Reviewer: Will they restart or will they have been rescheduled, so they stop?

Also, I would consider moving this entire paragraph somewhere near the top of this page, so that we have a clear description of the default behavior before describing how `disconnect` changes it.
